src/stage1_find_marks.cpp

```diff
@@ -139,13 +143,17 @@ uint64_t neonmovemask_bulk(uint8x16_t p0, uint8x16_t p1, uint8x16_t p2, uint8x16
 // a straightforward comparison of a mask against input. 5 uops; would be
 // cheaper in AVX512.
 really_inline uint64_t cmp_mask_against_input(simd_input in, uint8_t m) {
-#ifdef __AVX2__
-  const __m256i mask = _mm256_set1_epi8(m);
-  __m256i cmp_res_0 = _mm256_cmpeq_epi8(in.lo, mask);
-  uint64_t res_0 = static_cast<uint32_t>(_mm256_movemask_epi8(cmp_res_0));
```

Comment:
@lemire should I be doing static_cast in the expressions below?

Reply:
I am not sure exactly what you are asking about but let me guess... So a `_mm_movemask_epi8` returns an `int` with the lower 16 bits set (so it is positive). It is safe to write `uint32_t x = _mm_movemask_epi8` or `uint64_t x = _mm_movemask_epi8`because the cast from `int` to `uint32_t` or `uint64_t` is unambiguous.


It is trickier with `_mm256_movemask_epi8` because it returns an `int` that can be negative. A cast/promotion then becomes problematic.

My take is that you can, safely, use explicit casts, but you might be able to omit them if you'd like.

I, personally, do not frown on "unnecessary casts". The C/C++ languages like to have all sorts of automatic casts/promotions but that's not necessarily a great feature.